contracts/SquadController.sol

```diff
@@ -0,0 +1,323 @@
+        uint16 fee; // in basis points
+        uint256 purchasePrice;
+    }
+    mapping(bytes32 => Contribution) public contributions;
+
+    mapping(uint256 => bytes32) public validLicenses;
+
+    constructor(
+        address reserveToken,
+        address _tokenClaimCheck,
+        uint16 _networkFeeRate,
+        uint16 _maxNetworkFee,
+        address _treasury,
+        address _curve
+    ) public {
+        require(
+            _treasury != address(0),
+            "SquadController: zero treasury address"
+        );
+        require(_curve != address(0), "SquadController: zero Curve address");
+        require(
+            _maxNetworkFee <= 10000,
+            "SquadController: max network fee > 100%"
+        );
+        require(
+            _networkFeeRate <= _maxNetworkFee,
+            "SquadController: network gee > max"
+        );
+        networkFeeRate = _networkFeeRate;
+        maxNetworkFee = _maxNetworkFee;
+        treasury = _treasury;
+        tokenClaimCheck = TokenClaimCheck(_tokenClaimCheck);
+        tokenFactory = new ContinuousTokenFactory(reserveToken);
+        accounting = new Accounting();
+        curve = Curve(_curve);
+    }
+
+    event NewContribution(
+        address contributor,
+        bytes32 contributionId,
+        address beneficiary,
+        uint16 fee,
+        uint256 purchasePrice,
+        address continuousToken,
+        string metadata
+    );
+
+    function newContribution(
+        bytes32 contributionId,
+        address beneficiary,
+        uint16 fee,
+        uint256 purchasePrice,
+        string calldata name,
+        string calldata symbol,
+        string calldata metadata
+    ) external {
+        require(
+            !exists(contributionId),
+            "SquadController: contribution already exists"
+        );
+        require(
+            beneficiary != address(0),
+            "SquadController: zero beneficiary address"
+        );
+
+        tokenFactory.newContinuousToken(
+            contributionId,
+            name,
+            symbol,
+            address(curve)
+        );
+
+        contributions[contributionId] = Contribution(
+            beneficiary,
+            fee,
+            purchasePrice
+        );
+
+        address tokenAddress = address(
+            tokenFactory.tokenAddress(contributionId)
+        );
+
+        emit NewContribution(
+            msg.sender,
+            contributionId,
+            beneficiary,
+            fee,
+            purchasePrice,
+            tokenAddress,
+            metadata
+        );
+    }
+
+    event BuyLicense(
+        address buyer,
+        bytes32 contributionId,
+        uint256 licenseId,
+        string name,
+        uint256 amount,
+        uint256 price
+    );
+
+    function buyLicense(
+        bytes32 contributionId,
+        uint256 amount,
+        uint256 maxPrice
+    ) external mustExist(contributionId) {
+        ERC20 token = ERC20(tokenFactory.tokenAddress(contributionId));
```

Comment:
tokenFactory is storing an ERC20--can we have it return that directly?

Reply:
We can use the controller's `tokenAddress` function and possibly remove that function from `continuousTokenFactory` entirely.